File: InterviewPrepare/BestTimeToBuyAndSellStockWithTransactionFee.java
/**
 * LeetCode 714 Best Time to Buy and Sell Stock with Transaction Fee
 */
class Solution {
    public int maxProfit(int[] prices, int fee) {
        if (prices == null || prices.length == 0) {
            return 0;
        }

        int[][] dp = new int[prices.length][2];
        dp[0][0] = 0;
        dp[0][1] = -prices[0];
        for (int i = 1; i < prices.length; i++) {
            dp[i][0] = Math.max(dp[i - 1][0],dp[i - 1][1] + prices[i] - fee);
            dp[i][1] = Math.max(dp[i - 1][1],dp[i - 1][0] - prices[i]);
        }
        return dp[prices.length - 1][0];
    }
}

File: BestTimeToBuyAndSellStockWithTransactionFee.go
func maxProfit(prices []int, fee int) int {
	if prices == nil || len(prices) == 0 {
		return 0
	}
	cur := prices[0] + fee
	res := 0
	for i := 1;i < len(prices);i++ {
		if prices[i] + fee < cur {
			cur = prices[i] + fee
		} else if prices[i] > cur {
			res += prices[i] - cur
			cur = prices[i]
		}
	}
    return res
}
